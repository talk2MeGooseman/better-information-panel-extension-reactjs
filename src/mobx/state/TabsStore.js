import { observable, computed, action, makeObservable } from "mobx";
import TabModel from "../model/TabModel";
import { uuid } from "../../services/Utils";
import {
  queryPanelInformation
} from "../../services/Ebs";
import {
  DEFAULT_BODY_TEXT,
  ACTIVE_STEP_1,
  ACTIVE_STEP_2,
  ACTIVE_STEP_3,
  ACTIVE_STEP_4,
} from "../../services/constants";
import { path } from 'ramda';

export default class TabsStore {
  tabs = [];
  client = null;
  saveState = "";
  loadingState = "pending";
  activeStep = ACTIVE_STEP_1;
  videoOverlayHeight = '80vh';
  videoComponentVisibility = true;
  videoComponentTransparent = true;
  videoToggleImageUrl = '';
  videoToggleButtonPosition = '';
  tabIndex = 0;

  constructor() {
    makeObservable(this, {
      tabs: observable,
      client: observable,
      saveState: observable,
      loadingState: observable,
      activeStep: observable,
      videoOverlayHeight: observable,
      videoComponentVisibility: observable,
      videoComponentTransparent: observable,
      videoToggleImageUrl: observable,
      videoToggleButtonPosition: observable,
      tabIndex: observable,
      tabCount: computed,
      fetchTabs: action,
      updateTabs: action,
      saveTabs: action
    });
  }

  get tabCount() {
    return this.tabs.length;
  }

  setVideoComponentVisability(value) {
    this.saveState = '';
    this.videoComponentVisibility = value;
  }

  setVideoComponentTransparency(value) {
    this.saveState = '';
    this.videoComponentTransparent = value;
  }

  setVideoToggleImageUrl(value) {
    this.saveState = '';
    this.videoToggleImageUrl = value;
  }

  setVideoToggleButtonPosition(value) {
    this.saveState = '';
    this.videoToggleButtonPosition = value;
  }

  addTab() {
    this.saveState = '';

    if (this.activeStep === ACTIVE_STEP_1) {
      this.activeStep = ACTIVE_STEP_2;
    }

    this.tabs.push(TabModel.fromJS(this, {
      id: uuid(),
      title: 'Add Title',
      textColor: '#000000',
      bgColor: '#ffffff',
      body: DEFAULT_BODY_TEXT,
    }));
  }

  fetchTabs() {
    this.loadingState = "pending"
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", ({ data }) => {
        this.loadingState = "done"

        const result = path(['channel', 'bipConfig'], data)
        console.log(result)

        if (!result.tabs) {
          this.addTab();
          return;
        }

        this.activeStep = ACTIVE_STEP_4;

        // Check if the setting are defined from server, could be undefined
        // And screw shit up, use default state
        if (result.transparent !== undefined) {
          this.videoComponentTransparent = result.transparent;
        }
        if (result.visibility !== undefined) {
          this.videoComponentVisibility = result.visibility;
        }
        if (result.togglePosition !== undefined) {
          this.videoToggleButtonPosition = result.togglePosition;
        }

        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  updateTabs() {
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", result => {
        if (!result.tabs) {
          this.addTab();
          return;
        }
        this.activeStep = ACTIVE_STEP_4;
        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  saveTabs() {
    this.saveState = "pending"
      setPanelInformation(this.token, this.toJSON()).then(
        // inline created action
        action("fetchSuccess", result => {
          this.saveState = "done"
          this.activeStep = ACTIVE_STEP_4;
        }),
        // inline created action
        action("fetchError", error => {
          this.saveState = "error"
        })
      )
  }

  toJSON() {
    let jTabs = this.tabs.map((tab) => {
      return tab.toJSON();
    });

    return {
      tabs: jTabs,
      visibility: this.videoComponentVisibility,
      transparent: this.videoComponentTransparent,
      togglePosition: this.videoToggleButtonPosition,
    };
  }
}
