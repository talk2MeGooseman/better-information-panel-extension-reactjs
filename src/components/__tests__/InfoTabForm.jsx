import React, { Children } from "react";
import { createShallow, createMount} from '@material-ui/core/test-utils';


import InfoTabForm from "../InfoTabForm";
import TabsStore from "../../mobx/state/TabsStore";

describe('InfoTabForm', () => {
  let tabsStore, shallow, mount;

  afterEach(() => {
    mount.cleanUp();
  });

  beforeEach(() => {
    shallow = createShallow({
      dive: true
    });
    mount = createMount();
    tabsStore = new TabsStore();
  });

  it('matches default state snapshot', () => {
    const wrapper = mount(<InfoTabForm tabsStore={tabsStore} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('adds a tab on button press', () => {
    const wrapper = mount(<InfoTabForm tabsStore={tabsStore} />);
    expect(wrapper.find('FormControl').children().length).toEqual(1);
    expect(tabsStore.tabCount).toEqual(0);
    wrapper.find('Button').simulate('click');
    expect(wrapper.find('FormControl').children().length).toEqual(2);
    expect(tabsStore.tabCount).toEqual(1);
  });

});