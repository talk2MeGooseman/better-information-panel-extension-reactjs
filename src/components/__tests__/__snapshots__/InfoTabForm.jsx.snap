// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoTabForm matches default state snapshot 1`] = `
<WithStyles(InfoTabForm)
  tabsStore={
    TabsStore {
      "activeStep": 0,
      "loadingState": "pending",
      "saveState": "",
      "tabs": ObservableArray [],
      "token": undefined,
      "videoOverlayHeight": undefined,
    }
  }
>
  <InfoTabForm
    classes={
      Object {
        "button": "InfoTabForm-button-3",
        "container": "InfoTabForm-container-1",
        "formControl": "InfoTabForm-formControl-2",
      }
    }
    tabsStore={
      TabsStore {
        "activeStep": 0,
        "loadingState": "pending",
        "saveState": "",
        "tabs": ObservableArray [],
        "token": undefined,
        "videoOverlayHeight": undefined,
      }
    }
  >
    <div
      className="InfoTabForm-container-1"
    >
      <WithStyles(FormControl)>
        <FormControl
          classes={
            Object {
              "fullWidth": "MuiFormControl-fullWidth-7",
              "marginDense": "MuiFormControl-marginDense-6",
              "marginNormal": "MuiFormControl-marginNormal-5",
              "root": "MuiFormControl-root-4",
            }
          }
          component="div"
          disabled={false}
          error={false}
          fullWidth={false}
          margin="none"
          required={false}
        >
          <div
            className="MuiFormControl-root-4"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            <WithStyles(Button)
              aria-label="add"
              className="InfoTabForm-button-3"
              color="primary"
              disabled={false}
              onClick={[Function]}
              variant="fab"
            >
              <Button
                aria-label="add"
                className="InfoTabForm-button-3"
                classes={
                  Object {
                    "colorInherit": "MuiButton-colorInherit-13",
                    "disabled": "MuiButton-disabled-18",
                    "fab": "MuiButton-fab-19",
                    "flatPrimary": "MuiButton-flatPrimary-10",
                    "flatSecondary": "MuiButton-flatSecondary-11",
                    "focusVisible": "MuiButton-focusVisible-17",
                    "fullWidth": "MuiButton-fullWidth-23",
                    "label": "MuiButton-label-9",
                    "mini": "MuiButton-mini-20",
                    "outlined": "MuiButton-outlined-12",
                    "raised": "MuiButton-raised-14",
                    "raisedPrimary": "MuiButton-raisedPrimary-15",
                    "raisedSecondary": "MuiButton-raisedSecondary-16",
                    "root": "MuiButton-root-8",
                    "sizeLarge": "MuiButton-sizeLarge-22",
                    "sizeSmall": "MuiButton-sizeSmall-21",
                  }
                }
                color="primary"
                disableFocusRipple={false}
                disabled={false}
                fullWidth={false}
                mini={false}
                onClick={[Function]}
                size="medium"
                type="button"
                variant="fab"
              >
                <WithStyles(ButtonBase)
                  aria-label="add"
                  className="MuiButton-root-8 MuiButton-raised-14 MuiButton-fab-19 MuiButton-raisedPrimary-15 InfoTabForm-button-3"
                  disabled={false}
                  focusRipple={true}
                  focusVisibleClassName="MuiButton-focusVisible-17"
                  onClick={[Function]}
                  type="button"
                >
                  <ButtonBase
                    aria-label="add"
                    centerRipple={false}
                    className="MuiButton-root-8 MuiButton-raised-14 MuiButton-fab-19 MuiButton-raisedPrimary-15 InfoTabForm-button-3"
                    classes={
                      Object {
                        "disabled": "MuiButtonBase-disabled-25",
                        "focusVisible": "MuiButtonBase-focusVisible-26",
                        "root": "MuiButtonBase-root-24",
                      }
                    }
                    disableRipple={false}
                    disabled={false}
                    focusRipple={true}
                    focusVisibleClassName="MuiButton-focusVisible-17"
                    onClick={[Function]}
                    tabIndex="0"
                    type="button"
                  >
                    <button
                      aria-label="add"
                      className="MuiButtonBase-root-24 MuiButton-root-8 MuiButton-raised-14 MuiButton-fab-19 MuiButton-raisedPrimary-15 InfoTabForm-button-3"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex="0"
                      type="button"
                    >
                      <span
                        className="MuiButton-label-9"
                      >
                        <pure(Add)>
                          <Add>
                            <WithStyles(SvgIcon)>
                              <SvgIcon
                                classes={
                                  Object {
                                    "colorAction": "MuiSvgIcon-colorAction-30",
                                    "colorDisabled": "MuiSvgIcon-colorDisabled-32",
                                    "colorError": "MuiSvgIcon-colorError-31",
                                    "colorPrimary": "MuiSvgIcon-colorPrimary-28",
                                    "colorSecondary": "MuiSvgIcon-colorSecondary-29",
                                    "root": "MuiSvgIcon-root-27",
                                  }
                                }
                                color="inherit"
                                viewBox="0 0 24 24"
                              >
                                <svg
                                  aria-hidden="true"
                                  className="MuiSvgIcon-root-27"
                                  focusable="false"
                                  viewBox="0 0 24 24"
                                >
                                  <g>
                                    <path
                                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                    />
                                  </g>
                                </svg>
                              </SvgIcon>
                            </WithStyles(SvgIcon)>
                          </Add>
                        </pure(Add)>
                      </span>
                      <WithStyles(TouchRipple)
                        center={false}
                        innerRef={[Function]}
                      >
                        <TouchRipple
                          center={false}
                          classes={
                            Object {
                              "child": "MuiTouchRipple-child-37",
                              "childLeaving": "MuiTouchRipple-childLeaving-38",
                              "childPulsate": "MuiTouchRipple-childPulsate-39",
                              "ripple": "MuiTouchRipple-ripple-34",
                              "ripplePulsate": "MuiTouchRipple-ripplePulsate-36",
                              "rippleVisible": "MuiTouchRipple-rippleVisible-35",
                              "root": "MuiTouchRipple-root-33",
                            }
                          }
                        >
                          <TransitionGroup
                            childFactory={[Function]}
                            className="MuiTouchRipple-root-33"
                            component="span"
                            enter={true}
                            exit={true}
                          >
                            <span
                              className="MuiTouchRipple-root-33"
                            />
                          </TransitionGroup>
                        </TouchRipple>
                      </WithStyles(TouchRipple)>
                    </button>
                  </ButtonBase>
                </WithStyles(ButtonBase)>
              </Button>
            </WithStyles(Button)>
          </div>
        </FormControl>
      </WithStyles(FormControl)>
    </div>
  </InfoTabForm>
</WithStyles(InfoTabForm)>
`;
